[JSC] Add Air OptimizePairedLoadStore phase

Constellation opened the pull request:
#### 3632268ff8926089279b8c844d3d92a2cee78e69
<pre>
[JSC] Add Air OptimizePairedLoadStore phase
<a href="https://bugs.webkit.org/show_bug.cgi?id=258293">https://bugs.webkit.org/show_bug.cgi?id=258293</a>
rdar://111026259

Reviewed by Keith Miller.

This patch attempts to fuse multiple stores into StorePair when possible.
We do this optimization very lately: one of the major target of this optimization is stack spills with %fp base.
We would like to perform this optimization *after* register allocation is done. Also since this is block-local analysis,
we would like to perform it after Air SimplifyCFG phase is done.

We currently do not attempt to merge multiple loads into LoadPair. The reason is that we observed 20% regression
in particular tests (JetStream3/async-fs / JetStream3/sync-fs) because it is repeatedly performing `ldp`
against recently `stur`-ed address, which causes much longer latency than just doing `ldur`. Until we find a good way to
determine when we merge multiple loads into LoadPair, we only do StorePair merging.

For StorePair merging, we observed very specific regression in JetStream2/crypto-md5-SP only when perfoming stp ongoing call-args.
As a conservative first step, we do not perform merging when the base is %sp or %fp and when they are not spills. To make it work,
we annotate `spill` infomration in AirKind.

This offers 2~% progression in JetStream2/tsf-wasm&apos;s Runtime.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::loadPair32):
(JSC::MacroAssemblerARM64::loadPair64):
(JSC::MacroAssemblerARM64::storePair32):
(JSC::MacroAssemblerARM64::storePair64):
* Source/JavaScriptCore/b3/air/AirGenerate.cpp:
(JSC::B3::Air::prepareForGeneration):
* Source/JavaScriptCore/b3/air/AirLowerStackArgs.cpp:
(JSC::B3::Air::lowerStackArgs):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/b3/air/AirOptimizePairedLoadStore.cpp: Added.
(JSC::B3::Air::tryLoadPair):
(JSC::B3::Air::optimizePairedLoadStore):
* Source/JavaScriptCore/b3/air/AirOptimizePairedLoadStore.h: Added.

Canonical link: <a href="https://commits.webkit.org/265591@main">https://commits.webkit.org/265591@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2132c8c87be6b1146500f878c3ea5bd56eb18709

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11263 "13 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11473 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11807 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12914 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10741 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13852 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11458 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13650 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11424 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12312 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9525 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13326 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9605 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10202 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17395 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/9565 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10675 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10358 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13571 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/10687 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10782 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8866 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/11363 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9952 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3080 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14227 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/11682 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1279 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10635 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2851 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -1227,20 +1227,20 @@ class LowerToAir {
                 case Air::Move32:
-                    if (isValidForm(Store32, Arg::ZeroReg, dest.kind()) && dest.isValidForm(Move, Width32))
-                        return Inst(Store32, m_value, zeroReg(), dest);
+                    if (isValidForm(move.opcode, Arg::ZeroReg, dest.kind()) && dest.isValidForm(move.opcode, Width32))
```
Was `dest.isValidForm(Move, Width32)` just always failing before on 64bit?

Constellation replied:
Actually it succeeded because we do not care Width32 part on ARM64. But anyway, this is the right form.

kmiller68 commented on Source/JavaScriptCore/b3/air/AirKind.h:
```diff
@@ -75,7 +75,10 @@ struct Kind {
-    bool effects : 1;
+    bool effects : 1 { false };
+
+    // This is a marker where this instruction is generated for stack spilling.
```
Nit: This marks whether this instruction was generated for stack spilling.

Constellation replied:
Fixed.

kmiller68 commented on Source/JavaScriptCore/b3/air/AirOptimizePairedLoadStore.cpp:
```diff
@@ -0,0 +1,311 @@
+        unsigned index = block->size();
+        while (index--) {
+            Inst& inst = block->at(index);
+            if (inst.hasNonArgEffects())
+                continue;
```
Why add this? Do any moves have non-arg effects?

Constellation replied:
We do not want to merge if it has an effect, for example, Wasm store, which can have a trap.
Then, reordering / merging them changes the semantics since trap and store is observable.

kmiller68 replied:
Ah I should have read the implementation of `hasNonArgEffects()`, I agree.

kmiller68 commented on Source/JavaScriptCore/b3/air/AirOptimizePairedLoadStore.cpp:
```diff
@@ -0,0 +1,311 @@
+                    continue;
+
+                if ((inst.args[0].isGPTmp() || inst.args[0].isZeroReg()) && inst.args[1].isAddr() && inst.args[1].base() != inst.args[0]) {
+                    // sp & fp slots usage is particularly different for call args and spills.
```
Nit: sp & fp slot usage is, in particular, different for call args and spills.

Constellation replied:
Changed.

kmiller68 commented on Source/JavaScriptCore/b3/air/AirOptimizePairedLoadStore.cpp:
```diff
@@ -0,0 +1,311 @@
+
+        auto logFound = [&](const Inst& newInst) {
+            if (AirOptimizePairedLoadStoreInternal::verbose) {
+                dataLogLnIf(AirOptimizePairedLoadStoreInternal::verbose, "FOUND  ", inst, " ", target, " => ", newInst);
```
Nit: I think these can just be `dataLogLn(...)`

Constellation replied:
Changed.

kmiller68 commented on Source/JavaScriptCore/b3/air/AirOptimizePairedLoadStore.cpp:
```diff
@@ -0,0 +1,311 @@
+            if (AirOptimizePairedLoadStoreInternal::verbose) {
+                dataLogLnIf(AirOptimizePairedLoadStoreInternal::verbose, "FAILED ", inst, " ", target);
+                for (unsigned i = 1; i < count; ++i)
+                    dataLogLnIf(AirOptimizePairedLoadStoreInternal::verbose, "    ", block->at(current - i));
```
Ditto

Constellation replied:
Changed.

kmiller68 commented on Source/JavaScriptCore/b3/air/AirOptimizePairedLoadStore.cpp:
```diff
@@ -0,0 +1,311 @@
+        }
+
+        // Because str pimm only takes unsigned offset, we tend to pick stackPointerRegister based offsetting.
+        // But it is possible that framePointerRegister based offsetting can offer the benefit here.
```
Nit: the benefit => a benefit

Constellation replied:
Changed.

kmiller68 commented on Source/JavaScriptCore/b3/air/AirOptimizePairedLoadStore.cpp:
```diff
@@ -0,0 +1,311 @@
+            // If some instructions between the current and target have memory-load and memory-store,
+            // then we cannot merge them since reordering can change the results.
+            // But this is really pessimistic: if the base is the same to the current instruction, and if the offset
+            // is different from the current instruction, it is OK actually.
```
This comment seems outdated? Also, it seems like it should be ok to reorder 3 loads to make a loadPair. Maybe we should comment on that.

Constellation replied:
This part is only for StorePair (this function only do store merging), so I think this is correct.

kmiller68 replied:
Ah right, good point.

kmiller68 reviewed (approved):
r=me.

Constellation committed:
[JSC] Add Air OptimizePairedLoadStore phase
https://bugs.webkit.org/show_bug.cgi?id=258293
rdar://111026259

Reviewed by Keith Miller.

This patch attempts to fuse multiple stores into StorePair when possible.
We do this optimization very lately: one of the major target of this optimization is stack spills with %fp base.
We would like to perform this optimization *after* register allocation is done. Also since this is block-local analysis,
we would like to perform it after Air SimplifyCFG phase is done.

We currently do not attempt to merge multiple loads into LoadPair. The reason is that we observed 20% regression
in particular tests (JetStream3/async-fs / JetStream3/sync-fs) because it is repeatedly performing `ldp`
against recently `stur`-ed address, which causes much longer latency than just doing `ldur`. Until we find a good way to
determine when we merge multiple loads into LoadPair, we only do StorePair merging.

For StorePair merging, we observed very specific regression in JetStream2/crypto-md5-SP only when perfoming stp ongoing call-args.
As a conservative first step, we do not perform merging when the base is %sp or %fp and when they are not spills. To make it work,
we annotate `spill` infomration in AirKind.

This offers 2~% progression in JetStream2/tsf-wasm's Runtime.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::loadPair32):
(JSC::MacroAssemblerARM64::loadPair64):
(JSC::MacroAssemblerARM64::storePair32):
(JSC::MacroAssemblerARM64::storePair64):
* Source/JavaScriptCore/b3/air/AirGenerate.cpp:
(JSC::B3::Air::prepareForGeneration):
* Source/JavaScriptCore/b3/air/AirLowerStackArgs.cpp:
(JSC::B3::Air::lowerStackArgs):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/b3/air/AirOptimizePairedLoadStore.cpp: Added.
(JSC::B3::Air::tryLoadPair):
(JSC::B3::Air::optimizePairedLoadStore):
* Source/JavaScriptCore/b3/air/AirOptimizePairedLoadStore.h: Added.

Canonical link: https://commits.webkit.org/265591@main
